# Employee Attrition

Employee attrition, or turnover is a significant challenge for many organisations. It can lead to loss of institutional knowledge, decreased productivity, and increased hiring and training costs. By understanding the factors that contribute to attrition, organizations can implement strategies to retain their valuable employees. In this project, we will use the data obtained from kaggle on employee attrition to identify and analyze the key factors contributing to employee attrition and develop a predictive model for understanding and forecasting the nature of attrition in the company.

## Project Goals

* Identify the key factors driving employee attrition.
* Develop predictive models to anticipate future attrition.
* Propose recommendations to address attrition and improve employee retention.

**Problem**: High employee atttrition rates are negatively impacting the organization's productivity, morale, and overall perfomance.

**Research Question**: What factors are contributing to employee attrition and how can we address these issues to improve retention rates?

## Data Aquisition and Preparation

### 1. Import Python Libries 

We will begin our project by importing the Libraries that we will use for data cleaning, manipulation, analysis and visualization among other activities. 

In [1]:
# Import pandas for analyzing, cleaning, exploring and manipulating data
import pandas as pd

# Import numpy for numerical data analysis (working with Arrays)
import numpy as np

# Import Matplotlib for creating static, interactive, and animated visualizatios in Python
import matplotlib.pyplot as plt

# import Seaborn for high-level interface for drawing attractive and informative statistical graphics.
import seaborn as sns

# Set to display a maximum of 100 columns 
pd.set_option('display.max_columns', 100)

### 2. Import Data

We will then import the **Employee Attrition** dataset which we obtained from kaggle. We will assign our dataset to a variable `df`.

In [2]:
# load Employee Attrition dataset alias "df"
df = pd.read_csv(r"C:\Users\j\Documents\Data Analytics Projects\Employee_Attrition\Attrition data.csv")

# Display the first five records of our dataset
df.head()

EmployeeID  Age Attrition     BusinessTravel              Department  \
0           1   51        No      Travel_Rarely                   Sales   
1           2   31       Yes  Travel_Frequently  Research & Development   
2           3   32        No  Travel_Frequently  Research & Development   
3           4   38        No         Non-Travel  Research & Development   
4           5   32        No      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  Gender  \
0                 6          2  Life Sciences              1  Female   
1                10          1  Life Sciences              1  Female   
2                17          4          Other              1    Male   
3                 2          5  Life Sciences              1    Male   
4                10          1        Medical              1    Male   

   JobLevel                    JobRole MaritalStatus  MonthlyIncome  \
0         1  Healthcare Representative       Married         131160   
1         1         Research Scientist        Single          41890   
2         4            Sales Executive       Married         193280   
3         3            Human Resources       Married          83210   
4         1            Sales Executive        Single          23420   

   NumCompaniesWorked Over18  PercentSalaryHike  StandardHours  \
0                 1.0      Y                 11              8   
1                 0.0      Y                 23              8   
2                 1.0      Y                 15              8   
3                 3.0      Y                 11              8   
4                 4.0      Y                 12              8   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0                1.0                      6               1   
1                 1                6.0                      3               5   
2                 3                5.0                      2               5   
3                 3               13.0                      5               8   
4                 2                9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  EnvironmentSatisfaction  \
0                        0                     0                      3.0   
1                        1                     4                      3.0   
2                        0                     3                      2.0   
3                        7                     5                      4.0   
4                        0                     4                      4.0   

   JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
0              4.0              2.0               3                  3  
1              2.0              4.0               2                  4  
2              2.0              1.0               3                  3  
3              4.0              3.0               2                  3  
4              1.0              3.0               3                  3

From the output above we can see that our dataset is loaded successfully. Let us now display some of the characteristics of our dataset. 

In [3]:
# Display the critical information of our dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 29 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   EmployeeID               4410 non-null   int64  
 1   Age                      4410 non-null   int64  
 2   Attrition                4410 non-null   object 
 3   BusinessTravel           4410 non-null   object 
 4   Department               4410 non-null   object 
 5   DistanceFromHome         4410 non-null   int64  
 6   Education                4410 non-null   int64  
 7   EducationField           4410 non-null   object 
 8   EmployeeCount            4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

**Observations** <br>
From the above output, we have been able to make the following observations:
1. Our dataset is made up of a total of **29** columns and **4410** rows (records).
2. Our dataset has got data of different data types: Integers, Floats, and Strings (Objects).
3. We have observed some missing values in sime columns as indicated by the number of non-null. If in a column, the number of non-null is less than the number of records, it means there values that are missing.

### 3. Clean and Process Data
This is where incorrect, corrupted, incorrectly formated, duplicated, or incomplete data within a dataset are fixed or removed.

#### a. Renaming column names
When working with data in Pandas it is often necessary to rename the columns of a DataFrame to make them more meaningful and easier to work with. In our case, we will change the case for our column names from **CamelCase** to **snakecase** and separate column names with an underscore.

In [4]:
df.columns = df.columns.str.lower()
df.columns = ['employee_id', 'age', 'attrition', 'business_travel', 'department',
       'distance_from_home', 'education', 'education_field', 'employee_count',
       'gender', 'job_level', 'job_role', 'marital_status', 'monthly_income',
       'num_companies_worked', 'over_18', 'percent_salary_hike', 'standard_hours',
       'stock_option_level', 'total_working_years', 'training_times_last_year',
       'years_at_company', 'years_since_last_promotion', 'years_with_curr_manager',
       'environment_satisfaction', 'job_satisfaction', 'work_life_balance',
       'job_involvement', 'performance_rating']
df.columns

Index(['employee_id', 'age', 'attrition', 'business_travel', 'department',
       'distance_from_home', 'education', 'education_field', 'employee_count',
       'gender', 'job_level', 'job_role', 'marital_status', 'monthly_income',
       'num_companies_worked', 'over_18', 'percent_salary_hike',
       'standard_hours', 'stock_option_level', 'total_working_years',
       'training_times_last_year', 'years_at_company',
       'years_since_last_promotion', 'years_with_curr_manager',
       'environment_satisfaction', 'job_satisfaction', 'work_life_balance',
       'job_involvement', 'performance_rating'],
      dtype='object')

#### b. Handling Missing Values
Handling missing data allows for a more unbiased representation of the underlying patterns in the data. We will begin by checking for missing values in our dataset and determine which best possible way for dealing with them.

In [5]:
# Check for missing values in our dataset
df.isna().sum()

employee_id                    0
age                            0
attrition                      0
business_travel                0
department                     0
distance_from_home             0
education                      0
education_field                0
employee_count                 0
gender                         0
job_level                      0
job_role                       0
marital_status                 0
monthly_income                 0
num_companies_worked          19
over_18                        0
percent_salary_hike            0
standard_hours                 0
stock_option_level             0
total_working_years            9
training_times_last_year       0
years_at_company               0
years_since_last_promotion     0
years_with_curr_manager        0
environment_satisfaction      25
job_satisfaction              20
work_life_balance             38
job_involvement                0
performance_rating             0
dtype: int64

Our dataset has got missing values in the following columns.
* `num_companies_worked` = 19
* environment_satisfaction = 25
* `job_satisfaction` = 20
* work_life_balance = 38
* `total_working_year` = 9

The missing values are making up of approximately **2%** of our dataset. This is a very small figure and if we may choose to drop the affected rows, an impact is expected to be very minimal, if any, on the results of our analysis. However, we understand that every data point is a source of information and we will respect that aspect by not dropping rows or columns with missing data, instead we will impute the missing values with their column mean since all columns with missing values have numerical values.

In [6]:
# Imputing missing values in column "num_companies_worked" with mean
df['num_companies_worked'].fillna(df['num_companies_worked'].mean(), inplace = True)

# Imputing missing values in column "environment_satisfaction" with mean
df['environment_satisfaction'].fillna(df['environment_satisfaction'].mean(), inplace = True)

# Imputing missing values in column "job_satisfaction" with mean
df['job_satisfaction'].fillna(df['job_satisfaction'].mean(), inplace = True)

# Imputing missing values in column "work_life_balance" with mean
df['work_life_balance'].fillna(df['work_life_balance'].mean(), inplace = True)

# Imputing missing values in column "total_working_year" with mean
df['total_working_years'].fillna(df['total_working_years'].mean(), inplace = True)

df.isna().sum()

employee_id                   0
age                           0
attrition                     0
business_travel               0
department                    0
distance_from_home            0
education                     0
education_field               0
employee_count                0
gender                        0
job_level                     0
job_role                      0
marital_status                0
monthly_income                0
num_companies_worked          0
over_18                       0
percent_salary_hike           0
standard_hours                0
stock_option_level            0
total_working_years           0
training_times_last_year      0
years_at_company              0
years_since_last_promotion    0
years_with_curr_manager       0
environment_satisfaction      0
job_satisfaction              0
work_life_balance             0
job_involvement               0
performance_rating            0
dtype: int64

From the output above, we can see that we successfully fixed missing values as there are no more missing values availbale.

#### c. Handling Duplicates
Duplicated data occurs when storing the same data entries in the same data storage system, or accross multiple systems. This occurs accidentally. Let us begin by checking for duplicates from our dataset.

In [7]:
# Check for Duplicates
df.duplicated().sum()

0

We do not have duplicates in our dataset as indicated by the preceeding output.

#### d. Addressing Outliers
The first thing we have to do before we start addressing outliers is to detect if our dataset has got outliers. Outliers are data points that differ significantly from other observations. Let us use **pandas.describe()** function to determine the columns that are likely to have outliers. We will consider numeric columns only.

In [8]:
# Display Descriptive statistics for our dataset
df.describe()

employee_id          age  distance_from_home    education  \
count  4410.000000  4410.000000         4410.000000  4410.000000   
mean   2205.500000    36.923810            9.192517     2.912925   
std    1273.201673     9.133301            8.105026     1.023933   
min       1.000000    18.000000            1.000000     1.000000   
25%    1103.250000    30.000000            2.000000     2.000000   
50%    2205.500000    36.000000            7.000000     3.000000   
75%    3307.750000    43.000000           14.000000     4.000000   
max    4410.000000    60.000000           29.000000     5.000000   

       employee_count    job_level  monthly_income  num_companies_worked  \
count          4410.0  4410.000000     4410.000000           4410.000000   
mean              1.0     2.063946    65029.312925              2.694830   
std               0.0     1.106689    47068.888559              2.493497   
min               1.0     1.000000    10090.000000              0.000000   
25%               1.0     1.000000    29110.000000              1.000000   
50%               1.0     2.000000    49190.000000              2.000000   
75%               1.0     3.000000    83800.000000              4.000000   
max               1.0     5.000000   199990.000000              9.000000   

       percent_salary_hike  standard_hours  stock_option_level  \
count          4410.000000          4410.0         4410.000000   
mean             15.209524             8.0            0.793878   
std               3.659108             0.0            0.851883   
min              11.000000             8.0            0.000000   
25%              12.000000             8.0            0.000000   
50%              14.000000             8.0            1.000000   
75%              18.000000             8.0            1.000000   
max              25.000000             8.0            3.000000   

       total_working_years  training_times_last_year  years_at_company  \
count          4410.000000               4410.000000       4410.000000   
mean             11.279936                  2.799320          7.008163   
std               7.774275                  1.288978          6.125135   
min               0.000000                  0.000000          0.000000   
25%               6.000000                  2.000000          3.000000   
50%              10.000000                  3.000000          5.000000   
75%              15.000000                  3.000000          9.000000   
max              40.000000                  6.000000         40.000000   

       years_since_last_promotion  years_with_curr_manager  \
count                 4410.000000              4410.000000   
mean                     2.187755                 4.123129   
std                      3.221699                 3.567327   
min                      0.000000                 0.000000   
25%                      0.000000                 2.000000   
50%                      1.000000                 3.000000   
75%                      3.000000                 7.000000   
max                     15.000000                17.000000   

       environment_satisfaction  job_satisfaction  work_life_balance  \
count               4410.000000       4410.000000        4410.000000   
mean                   2.723603          2.728246           2.761436   
std                    1.089654          1.098753           0.703195   
min                    1.000000          1.000000           1.000000   
25%                    2.000000          2.000000           2.000000   
50%                    3.000000          3.000000           3.000000   
75%                    4.000000          4.000000           3.000000   
max                    4.000000          4.000000           4.000000   

       job_involvement  performance_rating  
count      4410.000000         4410.000000  
mean          2.729932            3.153741  
std           0.711400            0.360742  
min           1.000000            3.000000  
25%           